File: 5-LongestPalindromicSubstring/5-LongestPalindromicSubstring.py
# Last updated: 7/25/2026, 7:29:48 PM
class Solution:
    def expandAroundCenter(self, s:str, l: int, r: int) -> int:
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        return r - l - 1

    def longestPalindrome(self, s: str) -> str:
        start = 0
        end = 0

        for i in range(len(s)):
            odd = self.expandAroundCenter(s, i, i)
            even = self.expandAroundCenter(s, i, i + 1)

            maximumVal = max(even, odd)
            if maximumVal > end - start:
                start = i - (maximumVal- 1) // 2
                end = i + (maximumVal // 2)
        
        return s[start : end + 1]

        
        

File: 5-LongestPalindromicSubstring/5-LongestPalindromicSubstring.cpp
// Last updated: 7/25/2026, 7:29:51 PM
class Solution {
    int expandAroundCenter(const string& s, int left, int right) {
        while (left >= 0 and right < s.size() and s[left] == s[right]) {
            left--;
            right++;
        }
        return right-left-1;
    }
public:
    string longestPalindrome(string s) {
        if (s.empty()) return "";

        int start = 0;
        int end = 0;

        for (int i = 0; i < s.size(); i++) {
            int odd = expandAroundCenter(s, i, i);
            int even = expandAroundCenter(s, i, i + 1);

            int maxLen = max(odd, even);

            if (maxLen > end - start) {
                start = i - (maxLen - 1) / 2;
                end = i + (maxLen / 2);
            }
        }
        return s.substr(start, end - start + 1);
    }
};